File: recyclerview-paginated/src/com/hendraanggrian/recyclerview/paginated/PaginationAdapter.java
package com.hendraanggrian.recyclerview.paginated;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class PaginationAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int ITEM_VIEW_TYPE_LOADING = Integer.MAX_VALUE - 50; // Magic

    @NonNull private final RecyclerView.Adapter originalAdapter;
    @NonNull private final LoadingAdapter loadingAdapter;
    private boolean isDisplaying = true;

    public PaginationAdapter(@NonNull RecyclerView.Adapter adapter, @NonNull LoadingAdapter loadingAdapter) {
        this.originalAdapter = adapter;
        this.loadingAdapter = loadingAdapter;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM_VIEW_TYPE_LOADING) {
            return loadingAdapter.onCreateViewHolder(parent, viewType);
        } else {
            return originalAdapter.onCreateViewHolder(parent, viewType);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (isLoadingRow(position)) {
            loadingAdapter.onBindViewHolder(holder, position);
        } else {
            originalAdapter.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemCount() {
        return isDisplaying ? originalAdapter.getItemCount() + 1 : originalAdapter.getItemCount();
    }

    @Override
    public int getItemViewType(int position) {
        return isLoadingRow(position) ? ITEM_VIEW_TYPE_LOADING : originalAdapter.getItemViewType(position);
    }

    @Override
    public long getItemId(int position) {
        return isLoadingRow(position) ? RecyclerView.NO_ID : originalAdapter.getItemId(position);
    }

    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(hasStableIds);
        originalAdapter.setHasStableIds(hasStableIds);
    }

    @NonNull
    public RecyclerView.Adapter getOriginalAdapter() {
        return originalAdapter;
    }

    boolean isDisplaying() {
        return isDisplaying;
    }

    public void setDisplaying(boolean isDisplaying) {
        if (this.isDisplaying != isDisplaying) {
            this.isDisplaying = isDisplaying;
            notifyDataSetChanged();
        }
    }

    boolean isLoadingRow(int position) {
        return isDisplaying && position == getLoadingRowPosition();
    }

    private int getLoadingRowPosition() {
        return isDisplaying ? getItemCount() - 1 : -1;
    }
}

File: recyclerview-paginated/src/com/hendraanggrian/recyclerview/paginated/PaginationAdapter.kt
package com.hendraanggrian.recyclerview.paginated

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class PaginationAdapter(
        val actualAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>,
        val loadingAdapter: LoadingAdapter
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val ITEM_VIEW_TYPE_LOADING = Integer.MAX_VALUE - 50 // Magic
    }

    internal var isDisplaying = true
        set(value) {
            if (isDisplaying != value) {
                field = value
                notifyDataSetChanged()
            }
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            if (viewType == ITEM_VIEW_TYPE_LOADING) loadingAdapter.onCreateViewHolder(parent, viewType)
            else actualAdapter.onCreateViewHolder(parent, viewType)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int): Unit =
            if (isLoadingRow(position)) loadingAdapter.onBindViewHolder(holder, position)
            else actualAdapter.onBindViewHolder(holder, position)

    override fun getItemCount(): Int = if (isDisplaying) actualAdapter.itemCount + 1 else actualAdapter.itemCount

    override fun getItemViewType(position: Int): Int = if (isLoadingRow(position)) ITEM_VIEW_TYPE_LOADING else actualAdapter.getItemViewType(position)

    override fun getItemId(position: Int): Long = if (isLoadingRow(position)) RecyclerView.NO_ID else actualAdapter.getItemId(position)

    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(hasStableIds)
        actualAdapter.setHasStableIds(hasStableIds)
    }

    internal fun isLoadingRow(position: Int): Boolean = isDisplaying && position == loadingRowPosition

    private val loadingRowPosition: Int get() = if (isDisplaying) itemCount - 1 else -1
}